src/postprocess/refine.jl

```diff
@@ -0,0 +1,144 @@
+#     *Practical error bounds for properties in plane-wave electronic structure
+#     calculations* Preprint, 2021. [arXiv](https://arxiv.org/abs/2111.01470)
+
+"""
+Result of calling the [`refine_scfres`](@ref) function.
+- `basis`: Refinement basis, larger than the basis used to
+           run a first [`self_consistent_field`](@ref) computation.
+- `ψ`, `ρ`, `occupation`: Quantities from the scfres, transferred to the refinement basis
+                          and with virtual orbitals removed.
+- `δψ`, `δρ`: First order corrections to the wavefunctions and density.
+              The sign is such that the refined quantities are ψ - δψ and ρ - δρ.
+"""
+struct RefinementResult
+    basis :: PlaneWaveBasis
+    ψ
+    ρ
+    occupation
+    δψ
+    δρ
+end
+
+"""
+Transfer the result of an SCF to a larger basis set,
+and compute first order corrections ("refinements") to the wavefunctions and density.
+
+Returns a [`RefinementResult`](@ref) instance that can be used to refine quantities of interest,
+through [`refine_density`](@ref) and [`refine_forces`](@ref).
+"""
+function refine_scfres(scfres, basis_ref::PlaneWaveBasis{T}; ΩpK_tol,
+                       occ_threshold=default_occupation_threshold(T), kwargs...) where {T}
+    basis = scfres.basis
+
+    @assert basis.model.lattice == basis_ref.model.lattice
+    @assert length(basis.kpoints) == length(basis_ref.kpoints)
+    @assert all(basis.kpoints[ik].coordinate == basis_ref.kpoints[ik].coordinate
+                for ik in 1:length(basis.kpoints))
+
+    # Virtual orbitals must be removed
+    ψ, occ = select_occupied_orbitals(basis, scfres.ψ, scfres.occupation; threshold=occ_threshold)
+    ψr = transfer_blochwave(ψ, basis, basis_ref)
+    ρr = transfer_density(scfres.ρ, basis, basis_ref)
+    _, hamr = energy_hamiltonian(basis_ref, ψr, occ; ρ=ρr)
+
+    # Compute the residual R(P)
+    # This recomputes ρr internally, but it's only a minor inefficiency.
+    res = compute_projected_gradient(basis_ref, ψr, occ)
+
+    # Compute M^{-1} R(P), with M^{-1} defined in [CDKL2021]
+    P = [PreconditionerTPA(basis_ref, kpt) for kpt in basis_ref.kpoints]
+    map(zip(P, ψr)) do (Pk, ψk)
+        precondprep!(Pk, ψk)
+    end
+
+    function apply_M(φk, Pk, δφnk, n)
+        proj_tangent_kpt!(δφnk, φk)
+        δφnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δφnk
+        proj_tangent_kpt!(δφnk, φk)
+        δφnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δφnk
+        proj_tangent_kpt!(δφnk, φk)
+    end
+
+    function apply_inv_M(φk, Pk, δφnk, n)
+        proj_tangent_kpt!(δφnk, φk)
+        op(x) = apply_M(φk, Pk, x, n)
+        function f_ldiv!(x, y)
+            x .= proj_tangent_kpt(y, φk)
+            x ./= (Pk.mean_kin[n] .+ Pk.kin)
+            proj_tangent_kpt!(x, φk)
+        end
+        J = LinearMap{eltype(φk)}(op, size(δφnk, 1))
+        δφnk = IterativeSolvers.cg(J, δφnk, Pl=FunctionPreconditioner(f_ldiv!),
+                  verbose=false, reltol=0, abstol=1e-15)
```

Comment:
Here are also some tolerances we may want to tweak, respectively some output / convergence statistics we may want to access (let's discuss ...)

Reply:
Since in my tests the CG would always converge in 1 iteration, I changed `abstol` to `100*eps(T)` with max 20 iterations and a warning if that is not enough to converge. We might have to increase the constants once we start testing the method on more complex systems, TBD.